File: cses/06_tree_algorithms/2080_fixed-length-paths-i/solution.py
import sys


def main():
    # Đọc toàn bộ input một lần rồi chuyển sang int (một lần convert ở tầng C).
    data = sys.stdin.buffer.read().split()
    if not data:
        return
    ints = list(map(int, data))
    n = ints[0]
    k = ints[1]

    # Dựng danh sách kề (cây vô hướng).
    adj = [[] for _ in range(n + 1)]
    pos = 2
    for _ in range(n - 1):
        a = ints[pos]; b = ints[pos + 1]; pos += 2
        adj[a].append(b)
        adj[b].append(a)

    if n == 1:
        sys.stdout.buffer.write(b"0\n")
        return

    # DFS ITERATIVE từ root 1 (dùng stack tường minh, tránh RecursionError):
    # tính parent[] và danh sách preorder order[].
    parent = [0] * (n + 1)
    order = []
    order_append = order.append
    visited = bytearray(n + 1)
    st = [1]
    visited[1] = 1
    while st:
        u = st.pop()
        order_append(u)
        for v in adj[u]:
            if not visited[v]:
                visited[v] = 1
                parent[v] = u
                st.append(v)

    # height[v] = số cạnh của đường đi xuống dài nhất từ v.
    # heavy[v]  = con đạt được height đó (0 nếu v là lá).
    # Duyệt đảo preorder nên con luôn được xử lý trước cha.
    height = [0] * (n + 1)
    heavy = [0] * (n + 1)
    for u in reversed(order):
        p = parent[u]
        if p:
            h = height[u] + 1
            if h > height[p]:
                height[p] = h
                heavy[p] = u

    # Dựng danh sách light child (mọi con trừ heavy child) và memory pool.
    # Mỗi heavy chain (đỉnh đầu chuỗi .. lá sâu nhất) chiếm một đoạn liên tiếp
    # kích thước height[đầu]+1. Đỉnh v tại offset j có Dptr[v] = base + j, nên
    # cnt[v][i] = D[Dptr[v] + i] và Dptr[v] = Dptr[heavy[v]] - 1 (thừa kế O(1)).
    light = [None] * (n + 1)
    Dptr = [0] * (n + 1)
    pool = 0
    for v in order:
        p = parent[v]
        if v == 1 or heavy[p] != v:    # v là đỉnh đầu một heavy chain (light child)
            if v != 1:
                lp = light[p]
                if lp is None:
                    light[p] = [v]
                else:
                    lp.append(v)
            base = pool
            pool += height[v] + 1
            cur = v
            off = base
            while cur:
                Dptr[cur] = off
                off += 1
                cur = heavy[cur]
    D = [0] * (pool + 1)

    ans = 0
    base_idx = k - 1
    # Duyệt bottom-up (đảo preorder): mọi con của u đã xong trước u, và mảng cnt
    # của heavy child đã sẵn nằm chung vùng nhớ với u.
    for u in reversed(order):
        du = Dptr[u]
        hu = height[u]
        D[du] = 1                      # cnt[u][0] = u; đã thừa kế cnt của heavy child
        lu = light[u]
        if lu is None:
            if k <= hu:
                ans += D[du + k]
            continue
        for c in lu:                   # chỉ light child
            dc = Dptr[c]
            ch = height[c]
            # Pairing: đỉnh cách c là h thì cách u là h+1; ghép với một đỉnh đã
            # gộp trước cách u đúng idx = k-(h+1). Điều kiện idx >= 1 loại bỏ
            # chính u và giữ hai đầu mút đều nằm thực sự dưới u.
            for h in range(ch + 1):
                idx = base_idx - h
                if idx < 1:
                    break              # idx chỉ giảm khi h tăng
                if idx <= hu:
                    ans += D[dc + h] * D[du + idx]
            # Merge cnt[c] vào cnt[u] (dịch lên một khoảng cách).
            dd = du + 1
            for h in range(ch + 1):
                D[dd + h] += D[dc + h]
        # Đường đi thẳng đứng nhận u làm đầu mút, dài đúng k.
        if k <= hu:
            ans += D[du + k]

    sys.stdout.buffer.write((str(ans) + "\n").encode())


main()

File: cses/06_tree_algorithms/2080_fixed-length-paths-i/solution.cpp
#include <iostream>
#include <vector>

int main() {
    std::ios::sync_with_stdio(false);
    std::cin.tie(nullptr);

    int n, target_length;
    std::cin >> n >> target_length;

    // Đọc cây và dựng danh sách kề (đồ thị vô hướng).
    std::vector<std::vector<int>> graph(n + 1);
    for (int edge = 0; edge < n - 1; ++edge) {
        int first, second;
        std::cin >> first >> second;
        graph[first].push_back(second);
        graph[second].push_back(first);
    }

    // DFS ITERATIVE từ root 1 (dùng stack tường minh, tránh tràn stack khi n lớn):
    // tính parent[] và thứ tự preorder trong order[].
    std::vector<int> parent(n + 1, 0);
    std::vector<int> order;
    order.reserve(n);
    std::vector<int> stack{1};
    parent[1] = -1;
    while (!stack.empty()) {
        const int node = stack.back();
        stack.pop_back();
        order.push_back(node);
        for (const int next : graph[node]) {
            if (next == parent[node]) {
                continue;
            }
            parent[next] = node;
            stack.push_back(next);
        }
    }

    // height[node] = số cạnh của đường đi xuống dài nhất từ node.
    // heavy_child[node] = con đạt được height đó; con này sẽ được "thừa kế"
    // mảng đếm khoảng cách mà không phải sao chép (long-path decomposition).
    // Duyệt bottom-up (đảo preorder): con luôn được xử lý trước cha.
    std::vector<int> height(n + 1, 0);
    std::vector<int> heavy_child(n + 1, 0);
    for (auto iterator = order.rbegin(); iterator != order.rend(); ++iterator) {
        const int node = *iterator;
        const int ancestor = parent[node];
        if (ancestor > 0 && height[node] + 1 > height[ancestor]) {
            height[ancestor] = height[node] + 1;
            heavy_child[ancestor] = node;
        }
    }

    // Cấp phát memory pool: mỗi heavy chain chiếm một đoạn liên tiếp kích thước
    // height+1. Với cạnh heavy u->v ta đặt pointer[u] + 1 == pointer[v], nhờ đó
    // cnt[u][d] dùng lại đúng ô cnt[v][d-1] (thừa kế heavy child trong O(1)).
    std::vector<int> pointer(n + 1, 0);
    int pool_size = 0;
    for (const int node : order) {
        if (node != 1 && heavy_child[parent[node]] == node) {
            continue;
        }
        // node là đỉnh đầu một heavy chain; gán con trỏ dọc theo cả chuỗi.
        int current = node;
        int offset = pool_size;
        pool_size += height[node] + 1;
        while (current != 0) {
            pointer[current] = offset++;
            current = heavy_child[current];
        }
    }

    std::vector<long long> distance_count(pool_size + 1, 0);
    long long answer = 0;
    const int complement_base = target_length - 1;

    // Duyệt bottom-up: khi xử lý node thì mọi con đã xong, riêng heavy child đã
    // sẵn nằm chung vùng nhớ với node.
    for (auto iterator = order.rbegin(); iterator != order.rend(); ++iterator) {
        const int node = *iterator;
        const int node_pointer = pointer[node];
        distance_count[node_pointer] = 1;   // cnt[node][0] = 1 (chính node)

        // Chỉ lặp qua các light child (mọi con trừ heavy child).
        for (const int child : graph[node]) {
            if (parent[child] != node || child == heavy_child[node]) {
                continue;
            }

            const int child_pointer = pointer[child];
            // Pairing: đỉnh cách child là child_depth thì cách node là
            // child_depth+1; ghép với một đỉnh đã gộp trước cách node đúng
            // previous_depth = k-(child_depth+1). Ràng buộc >= 1 loại bỏ chính
            // node và bảo đảm hai đầu mút đều nằm thực sự dưới node.
            for (int child_depth = 0;
                 child_depth <= height[child]; ++child_depth) {
                const int previous_depth = complement_base - child_depth;
                if (previous_depth < 1) {
                    break;
                }
                if (previous_depth <= height[node]) {
                    answer += distance_count[child_pointer + child_depth] *
                              distance_count[node_pointer + previous_depth];
                }
            }

            // Merge cnt[child] vào cnt[node] (dịch lên một khoảng cách).
            for (int child_depth = 0;
                 child_depth <= height[child]; ++child_depth) {
                distance_count[node_pointer + child_depth + 1] +=
                    distance_count[child_pointer + child_depth];
            }
        }

        // Trường hợp node là một đầu mút của đường đi thẳng đứng dài đúng k.
        if (target_length <= height[node]) {
            answer += distance_count[node_pointer + target_length];
        }
    }

    std::cout << answer << '\n';
    return 0;
}
